File: python/problems/leetcode/0279_perfect_squares/solution.py
"""LeetCode 279: Perfect Squares."""

from __future__ import annotations


class Solution:
    def numSquares(self, n: int) -> int:
        dp = [0] + [n + 1] * n
        for i in range(1, n + 1):
            j = 1
            while j * j <= i:
                if dp[i - j * j] + 1 < dp[i]:
                    dp[i] = dp[i - j * j] + 1
                j += 1
        return dp[n]


if __name__ == "__main__":
    assert Solution().numSquares(12) == 3
    assert Solution().numSquares(13) == 2
    assert Solution().numSquares(1) == 1
    print("0279 OK")

File: cpp/problems/leetcode/0279_perfect_squares/solution.cpp
// LeetCode 279
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int numSquares(int n) {
        vector<int> dp(n + 1, n + 1);
        dp[0] = 0;
        for (int i = 1; i <= n; ++i)
            for (int j = 1; j * j <= i; ++j)
                dp[i] = min(dp[i], dp[i - j * j] + 1);
        return dp[n];
    }
};

int main() {
    assert(Solution().numSquares(12) == 3);
    assert(Solution().numSquares(13) == 2);
    assert(Solution().numSquares(1) == 1);
    cout << "0279 OK" << endl;
    return 0;
}
